### ISSUE TITLE
highlight on wildcard field type does not produce result

### ISSUE THREAD

**Issue by ragnarok56** [elastic/elasticsearch#96013]
### Elasticsearch Version

8.7.1

### Java Version

_bundled_

### OS Version

Centos 7.9

### Problem Description

When querying an index containing a field with a `wildcard` field type, attempting to highlight the matching text from that field does not return any matches for the `wildcard` field, but other highlights work, including those for `keyword` field types.  I attempted a variety of highlighter settings and configurations to see if there was some trick to get it to work, but the only time it ever showed up was when i set `no_match_size` to a larger number than 0, which kind of defeated the purpose since it then shows up in every query.

I couldn't find any other examples of this being an issue except in this forum topic (https://discuss.elastic.co/t/highlighting-on-new-wildcard-field/305525) where the answer was to just roll your own highlighting for `wildcard` fields.  I haven't been able to find mention of this being a limitation in elasticsearch docs for either highlighting or the `wildcard` field type in particular.  `wildcard` is also listed under the `keyword` family of field types, so we expect this to work with highlighting.

### Steps to Reproduce

Create index with keyword and wildcard field types
```
PUT example
{
  "mappings": {
    "properties": {
      "wildcard_field": {
        "type": "wildcard"
      }, 
      "keyword_field": {
        "type": "keyword"
      }
    }
  }
}
```
add doc with value for each
```
POST example/_doc
{
  "wildcard_field": "test wildcard value",
  "keyword_field": "test keyword value"
}
```
query for both and expect highlights for both
```
GET example/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "wildcard": {
            "wildcard_field": {
              "value": "*wildcard*"
            }
          }
        },
        {
          "wildcard": {
            "keyword_field": {
              "value": "*keyword*"
            }
          }
        }
      ]
    }
  },
  "highlight": {
    "fields": {
      "wildcard_field": {},
      "keyword_field": {}
    }
  }
}
```

result only contains highlight for keyword type and not wildcard type
```
{
  "took": 22,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 2,
    "hits": [
      {
        "_index": "example",
        "_id": "IDBhCIgBL3-lbiIFygte",
        "_score": 2,
        "_source": {
          "wildcard_field": "test wildcard value",
          "keyword_field": "test keyword value"
        },
        "highlight": {
          "keyword_field": [
            "<em>test keyword value</em>"
          ]
        }
      }
    ]
  }
}
```

**Comment by thecoop** [comment]
@ragnarok56 I found [this issue](https://github.com/elastic/elasticsearch/issues/14943) from 2015 which looks similar. This may not be a scenario that is supported (in which case we should improve the docs around this). But the search team will be able to confirm.

**Comment by ragnarok56** [comment]
I don't think its the wildcard query that is failing to produce the highlight, in my example that actually does what I expected for the keyword field.  It's that the wildcard mapping type, using any type of query at all, never produces a highlight.

**Comment by bogdanul2003** [comment]
Same issue in older versions as well. Works when using 
"require_field_match" : false
The problem is that if you add other fields to highlight that you didn't searched for in the query, those will be included in the highlights if there was a match.

`GET example/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "wildcard": {
            "keyword_field": {
              "value": "*keyword*"
            }
          }
        }
      ]
    }
  },
  "highlight": {
    "require_field_match": "true",
    "fields": {
      "wildcard_field": {},
      "keyword_field": {}
    }
  }
}
 `

**Comment by coderliguoqing** [comment]
yeah,  I also encountered the same problem, when I searched for a wildcard field type and needed to highlight it for return, it did not return anything, but everything was normal on non wildcard fields；
example:
`GET /my_index/_search
{
  "query": {
    "wildcard": {
      "flowId": "*2412230132*"
    }
  },
  "highlight": {
    "fields": {
      "*": {}
    }
  }
}`
mapping:
`PUT my_index
{
  "mappings": {
    "properties": {
      "flowId": {
        "type": "wildcard"
      },
    }
  }
}`

**Comment by coderliguoqing** [comment]
Is there a plan to fix this issue in which version？ @elasticsearchmachine

**Comment by benwtrent** [comment]
What would the expected result for the wildcard highlighting be? Simply highlighting the entire text field? 

This is not on plan to be fixed immediately, and is a new enhancement as we have purposefully not implemented highlighting in this way over wildcard fields.

**Comment by coderliguoqing** [comment]
> What would the expected result for the wildcard highlighting be? Simply highlighting the entire text field?通配符高亮显示的预期结果是什么？只是突出显示整个文本字段？
> 
> This is not on plan to be fixed immediately, and is a new enhancement as we have purposefully not implemented highlighting in this way over wildcard fields.这不是立即修复的计划，而是一项新的增强功能，因为我们故意没有以这种方式在通配符字段上实现高亮显示。

I think the expected result is that the content information matched by the wildcard should be highlighted, for example: { "query": { "wildcard": { "flowId": "*2412230132*" } }, "highlight": { "fields": { "*": {} } }, then in the returned highlight, it should contain flowId and the highlighted part is 2412230132, This is the same as any other keyword such as match; This should be a generic resultant ability, and it should not be distinguished that he is a wildcard but does not take effect
